File: bk2ch16p687AVQueuePlayer/ch29p952AVQueuePlayer/Player.m


#import "Player.h"
@import AVFoundation;
@import MediaPlayer;

@interface Player () <AVAudioPlayerDelegate>
@end

@implementation Player

- (void) play: (NSURL*) path {
    NSURL *fileURL = path;
    NSError* err = nil;
    AVAudioPlayer *newPlayer =
        [[AVAudioPlayer alloc] initWithContentsOfURL: fileURL error: &err];
    if (!newPlayer) {
        NSLog(@"%@", err);
        return;
    }

    self.player = newPlayer; // retain policy

    [self.player prepareToPlay];
    [self.player setDelegate: self];
    if (self.forever)
        self.player.numberOfLoops = -1;
    [self.player play];
    
    // cute little demo
    MPNowPlayingInfoCenter* mpic = [MPNowPlayingInfoCenter defaultCenter];
    mpic.nowPlayingInfo =
    @{MPMediaItemPropertyTitle:@"This Is a Test"};

}

- (void)audioPlayerDidFinishPlaying:(AVAudioPlayer *)player // delegate method
                       successfully:(BOOL)flag {
    [[NSNotificationCenter defaultCenter] 
     postNotificationName:@"soundFinished" object:nil];
}

// unnecessary to implement audioPlayerEndInterruption delegate method...
// ... merely to reactivate audio session after interrupt; happens automatically
// also, automatically resumes after interruption


-(void)audioPlayerBeginInterruption:(AVAudioPlayer *)player {
    NSLog(@"audio player interrupted");
}

-(void)audioPlayerEndInterruption:(AVAudioPlayer *)player withOptions:(NSUInteger)opts {
    NSLog(@"audio player ended interruption options: %lu", (unsigned long)opts);
}


@end

File: bk2ch16p687AVQueuePlayer/ch29p952AVQueuePlayer/Player.swift


import UIKit
import AVFoundation
import MediaPlayer

protocol PlayerDelegate : class {
    func soundFinished(sender : AnyObject)
}

class Player : NSObject, AVAudioPlayerDelegate
{
    var player : AVAudioPlayer!
    var forever = false
    weak var delegate : PlayerDelegate?

    func playFileAtURL(fileURL:NSURL) {
        player?.delegate = nil
        player?.stop()
        player = try! AVAudioPlayer(contentsOfURL: fileURL)
        // error-checking omitted
        player.prepareToPlay()
        player.delegate = self
        if self.forever {
            self.player.numberOfLoops = -1
        }
        player.play()
        
    }
    
    // delegate method
    func audioPlayerDidFinishPlaying(player: AVAudioPlayer, successfully flag: Bool) {
        delegate?.soundFinished(self)
    }
        
    func stop () {
        player?.pause()
    }

    deinit {
        player?.delegate = nil
    }


}
